lua/rip-substitute/matches.lua

```diff
@@ -0,0 +1,221 @@
+		for i, match in ipairs(matches) do
+			local rgArgs = {
+				"--line-number",
+				"--column",
+				"--vimgrep",
+				"--no-filename",
+				"--only-matching",
+				"--replace=" .. toReplace,
+				toSearch,
+			}
+			local replaced
+			code, replaced = rg.runRipgrep(rgArgs)
+			if code ~= 0 then
+				return nil, "[3]could not get rg replacements"
+			end
+			if #matched ~= #replaced then return nil, "#matched ~= #replaced" end
+			if not replaced[i] then
+				return nil, "[4]could not get replacement for " .. match.matchedText
+			end
+
+			local line = replaced[i]
+			local rowStr, colStr, replacementText = line:match("^(%d+):(%d+):(.*)")
+			---@type RipSubstituteMatch
+			match.replacementText = replacementText
+		end
+	end
+	return matches
+end
+
+---@param matches RipSubstituteMatch[]
+---@return RipSubstituteMatch | nil, string |nil
+function M.getClosestMatchAfterCursor(matches)
+	if #matches == 0 then return nil end
+	local state = require("rip-substitute.state").state
+	local cursor_row, cursor_col = unpack(vim.api.nvim_win_get_cursor(state.targetWin))
+	local closestMatch = nil -- Store the closest match found after cursor
+
+	cursor_row = cursor_row - 1
+	-- First, try to find a match after the cursor position
+	for _, match in ipairs(matches) do
+		local on_line_after = match.row > cursor_row
+		local on_same_line = match.row == cursor_row
+		local cursor_on_match = on_same_line and match.col <= cursor_col and match.col >= cursor_col
+		local on_same_line_after = not cursor_on_match and on_same_line and match.col > cursor_col
+
+		if on_line_after or cursor_on_match or on_same_line_after then
+			closestMatch = match
+			break -- Stop the loop if a match is found
+		end
+	end
+
+	-- If no match is found after cursor, search from the beginning of the file to the cursor
+	if not closestMatch then
+		for _, match in ipairs(matches) do
+			local on_line_before = match.row < cursor_row
+			local on_same_line_before = match.row == cursor_row and match.col < cursor_col
+
+			if on_line_before or on_same_line_before then
+				closestMatch = match
+				-- No break here; keep updating closestMatch to the last match before cursor
+			end
+		end
+	end
+
+	return closestMatch
+end
+
+---@param replacing boolean
+---@param newSelectedMatchIndex number
+local function updateSelectedMatchHighlight(replacing, newSelectedMatchIndex)
+	local state = require("rip-substitute.state").state
+	vim.api.nvim_buf_clear_namespace(
+		state.targetBuf,
+		state.incPreviewNs,
+		state.selectedMatch.row,
+		state.selectedMatch.row + 1
+	)
+	if replacing then
+		rg.highlightReplacement(state.selectedMatch, false, state.targetBuf, state.incPreviewNs)
+	else
+		rg.highlightMatch(
+			state.selectedMatch,
+			false,
+			state.targetBuf,
+			state.incPreviewNs,
+			state.selectedMatch.col + #state.selectedMatch.matchedText
+		)
+	end
+	state.selectedMatch = state.matches[newSelectedMatchIndex]
+	--TODO: this will probably clear too many highlights
```

Comment:
yes, an issue will be multiple matches in one line, for instance. Not totally sure what would be the best solution here though